vllm/spec_decode/ngram_worker.py

```diff
@@ -0,0 +1,260 @@
+import copy
+from typing import Dict, List, Optional
+
+import numpy as np
+import torch
+
+from vllm.sequence import SamplerOutput, SequenceGroupMetadata
+from vllm.spec_decode.interfaces import SpeculativeProposals
+from vllm.spec_decode.top1_proposer import DraftModelTop1Proposer
+from vllm.worker.worker import Worker
+
+
+class NGramWorker(Worker):
```

Comment:
Let's make this inherit from WorkerBase, not Worker. Unless I miss something we don't need anything from the Worker class.

Reply:
I try use the WorkerBase here, but got below error:
TypeError: Can't instantiate abstract class NGramWorker with abstract methods add_lora, determine_num_available_blocks, execute_model, get_cache_block_size_bytes, initialize_cache, list_loras, remove_lora

Reply:
> Got it, shall we also need make multi_step_worker inherit from WorkerBase too?

multi_step_worker should ideally inherit from WorkerBase, but it requires a larger refactor. You can help with that if you want :) the TL;DR is to decouple the sampler from CPU serialization of python objects. If we do this then the multi step worker can inherit directly WorkerBase and become decoupled from Worker. But it's a lot of work right now, with the only benefit being cleaner software design. We'll do it soon as it will also improve performance, but not now.

> I try use the WorkerBase here, but got below error:
TypeError: Can't instantiate abstract class NGramWorker with abstract methods add_lora, determine_num_available_blocks, execute_model, get_cache_block_size_bytes, initialize_cache, list_loras, remove_lora

You can inherit from [`LoraNotSupportedWorkerBase`](https://github.com/vllm-project/vllm/blob/1e8f4252aa163041094a8fedbb701a98d7087d7c/vllm/worker/worker_base.py#L81-L93) to handle the lora methods. `determine_num_available_blocks` can be a noop; [the proposer worker's `determine_num_available_blocks` is unused in spec decode](https://github.com/vllm-project/vllm/blob/1e8f4252aa163041094a8fedbb701a98d7087d7c/vllm/spec_decode/spec_decode_worker.py#L140-L159). For `get_cache_block_size_bytes` you can return 0, since there is no on-device KV cache for ngram. Initialize_cache can be noop as well (just `pass`), unless there's on-device state you want to initialize.

Reply:
Got it, would apply it in latest commit